README.md
# HTagView

<!--[![CI Status](http://img.shields.io/travis/Chang, Hao/HTagView.svg?style=flat)](https://travis-ci.org/Chang, Hao/HTagView)-->
[![Version](https://img.shields.io/cocoapods/v/HTagView.svg?style=flat)](http://cocoapods.org/pods/HTagView)
[![License](https://img.shields.io/cocoapods/l/HTagView.svg?style=flat)](http://cocoapods.org/pods/HTagView)
[![Platform](https://img.shields.io/cocoapods/p/HTagView.svg?style=flat)](http://cocoapods.org/pods/HTagView)

HTagView is a customized tag view sublassing UIView where tag could be either with cancel button or multiseletable.

### Features

- `.cancel` and `.select` tag types available (see below)
- Single / Multi selection for `.select` tags
- `HTagViewDataSource` and `HTagViewDelegte` protocols
- Customized configuration
- Supporting `@IBDesignable` and autolayout
- Specific/Auto tag width
- Specific/Auto Maximum tag width

### Demo
![](demo.gif)




## Example

To run the example project, clone the repo, and run `pod install` from the Example directory first.


## Installation

HTagView is available through [CocoaPods](http://cocoapods.org). To install
it, simply add the following line to your Podfile:

```ruby
pod "HTagView"
```

### Requirements

Swift Version | HTagView Version
----- | -----
Swift 2.3 | 2.0.x
Swift 3.0 | 2.1.0 or later

## Usage
### Configure HTagView
These properties support `@IBDesignable` as well.

```swift
override func viewDidLoad(){
	super.viewDidLoad()
	//.
	//.
	//.

	// configure tagView
	tagView.delegate = self
	tagView.dataSource = self
	tagView.marg = 20
	tagView.btwTags = 20
	tagView.btwLines = 20
	tagView.tagFont = UIFont.systemFont(ofSize: 15)
	tagView.tagMainBackColor = UIColor(red: 1, green: 130/255, blue: 103/255, alpha: 1)
	tagView.tagSecondBackColor = UIColor.lightGray
	tagView.tagSecondTextColor = UIColor.darkText
	tagView.tagContentEdgeInsets = UIEdgeInsets(top: 10, left: 20, bottom: 10, right: 20)
	tagView.tagMaximumWidth = .HTagAutoMaximumWidth
	tagView.tagBorderColor = UIColor.black.cgColor
	tagView.tagBorderWidth = 2
}
```
### Data Source
```swift
class ViewController: UIViewController, HTagViewDataSource{
	// .
	// .
	// .

    // MARK: - Data
    let data = ["Hey!","This","is","a","HTagView."]

    
    // MARK: - HTagViewDataSource
    func numberOfTags(_ tagView: HTagView) -> Int {
    	return data.count
    }
    
    func tagView(_ tagView: HTagView, titleOfTagAtIndex index: Int) -> String {
        return data[index]
    }
    
    func tagView(_ tagView: HTagView, tagTypeAtIndex index: Int) -> HTagType {
		return .select
		// return .cancel
    }
    
    func tagView(_ tagView: HTagView, tagWidthAtIndex index: Int) -> CGFloat {
        return .HTagAutoWidth
        // return 150
    }
}
```
### Delegate
```swift
class ViewController: UIViewController, HTagViewDelegate, HTagViewDataSource {
	// .
	// .
	// .

    // MARK: - HTagViewDelegate
    func tagView(_ tagView: HTagView, tagSelectionDidChange selectedIndices: [Int]) {
        print("tag with indices \(selectedIndices) are selected")
    }
    func tagView(_ tagView: HTagView, didCancelTagAtIndex index: Int) {
        print("tag with index: '\(index)' has to be removed from tagView")
        data.remove(at: index)
        tagView.reloadData()
    }
}
```
### Manually select/deselect tags
```swift
tagView.selectTagAtIndex(6)
tagView.deselectTagAtIndex(3)
```

## Author

Hao


## License

HTagView is available under the MIT license. See the LICENSE file for more info.


CHANGELOG.md
# Change Log

### 3.0.1
- add default `.HTagAutoWidth` as default implementation
- silent tapped and cancelled log

### 3.0
- rebuild `HTag` class
- add tag width configuration
	- auto / specified width
	- auto / specified maximum width

### 2.1.x
- Swift 3.0

### 2.0.x
- Swift 2.3
- Single selection feature
- Better implementation of MVC design patter
	- `HTagViewDataSource`
	- `HTagViewDelegate`
- Huge change in APIs. (For more information, please checkout document)

### 1.1.2
- Update APIs

	- `setTagsWithTitles(_:)`
	- `selectTagWithTitles(_:)`
	- `deselectTagWithTitles(_:)`
	- `addTagWithTitle(_:)`
	- `tagView.removeTagWithTitle(:)`
- Documentation


### 1.1.1
- First stable version.

Example/Tests/Tests.swift
// https://github.com/Quick/Quick
/*
import Quick
import Nimble
import HTagView

class TableOfContentsSpec: QuickSpec {
    override func spec() {
        describe("these will fail") {

            it("can do maths") {
                expect(1) == 2
            }

            it("can read") {
                expect("number") == "string"
            }

            it("will eventually fail") {
                expect("time").toEventually( equal("done") )
            }
            
            context("these will pass") {

                it("can do maths") {
                    expect(23) == 23
                }

                it("can read") {
                    expect("🐮") == "🐮"
                }

                it("will eventually pass") {
                    var time = "passing"

                    DispatchQueue.main.async {
                        time = "done"
                    }

                    waitUntil { done in
                        Thread.sleep(forTimeInterval: 0.5)
                        expect(time) == "done"

                        done()
                    }
                }
            }
        }
    }
}*/


HTagView/Classes/Tag.swift
//
//  Tag.swift
//  Pods
//
//  Created by Chang, Hao on 12/12/2016.
//
//

import UIKit

protocol TagDelegate: class {
    func tagClicked(_ sender: Tag)
    func tagCancelled(_ sender: Tag)
}

protocol TagDataSource: class{
    
    
}

public class Tag: UIView {
    weak var delegate: TagDelegate?
    
    // MARK: - HTag Configuration
    /**
     Type of tag
     */
    var tagType: HTagType = .cancel { didSet { updateAll() } }
    /**
     Title of tag
     */
    var tagTitle: String = "tag" { didSet { updateAll() } }
    /**
     Main background color of tags
     */
    var tagMainBackColor : UIColor = UIColor.white { didSet { updateTitlesColorsAndFontsDueToSelection() } }
    /**
     Main text color of tags
     */
    var tagMainTextColor : UIColor = UIColor.black { didSet { updateTitlesColorsAndFontsDueToSelection() } }
    /**
     Secondary background color of tags
     */
    var tagSecondBackColor : UIColor = UIColor.gray { didSet { updateTitlesColorsAndFontsDueToSelection() } }
    /**
     Secondary text color of tags
     */
    var tagSecondTextColor : UIColor = UIColor.white { didSet { updateTitlesColorsAndFontsDueToSelection() } }
    /**
     The border width to height ratio of HTags.
     */
    var tagBorderWidth :CGFloat = 1 { didSet { updateBorder() } }
    /**
     The border color to height ratio of HTags.
     */
    var tagBorderColor :CGColor? = UIColor.darkGray.cgColor { didSet { updateBorder() } }
    
    /**
     The corner radius to height ratio of HTags.
     */
    var tagCornerRadiusToHeightRatio :CGFloat = 0.2 { didSet { updateBorder() } }
    /**
     The content EdgeInsets of HTags, which would automatically adjust the position in `.cancel` type.
     On the other word, usages are the same in both types.
     */
    var tagContentEdgeInsets: UIEdgeInsets = UIEdgeInsets() { didSet { updateAll()} }
    /**
     The Font of HTags.
     */
    var tagFont: UIFont = UIFont.systemFont(ofSize: 17) { didSet { updateAll() } }
    
    // MARK: - status
    private(set) var isSelected: Bool = false
    
    
    // MARK: - Subviews
    var button: UIButton = UIButton(type: .system)
    var cancelButton: UIButton = UIButton(type: .system)
    
    // MARK: - Init
    override init(frame: CGRect){
        super.init(frame: frame)
        configure()
    }
    
    required public init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
        configure()
    }
    
    deinit {
        button.removeObserver(self, forKeyPath: "highlighted")
    }
    
    // MARK: - Configure
    func configure(){
        setupButton()
        setupCancelButton()
        updateAll()
    }
    
    // MARK: - button
    func setupButton() {
        addSubview(button)
        button.addTarget(self, action: #selector(tapped), for: .touchUpInside)
        button.addObserver(self, forKeyPath: "highlighted", options: .new, context: nil)
    }
    
    func setHighlight(_ highlighted: Bool) {
        let color = isSelected ? tagMainBackColor : tagSecondBackColor
        backgroundColor = highlighted ? color.darker() : color
    }
    func setSelected(_ selected: Bool) {
        isSelected = selected
        updateTitlesColorsAndFontsDueToSelection()
    }
    public override func observeValue(forKeyPath keyPath: String?, of object: Any?, change: [NSKeyValueChangeKey : Any]?, context: UnsafeMutableRawPointer?) {
        
        guard let keyPath = keyPath else {
            return
        }
        switch keyPath {
        case "highlighted":
            guard let highlighted = change?[.newKey] as? Bool else  {
                break
            }
            setHighlight(highlighted)
        default:
            break
        }
    }
    
    // MARK: - cancel button
    func setupCancelButton() {
        cancelButton.setImage(UIImage(named: "close_small", in: Bundle(for: self.classForCoder), compatibleWith: nil)?.withRenderingMode(.alwaysTemplate), for: UIControlState())
        cancelButton.addTarget(self, action: #selector(cancelled), for: .touchUpInside)
        addSubview(cancelButton)
        cancelButton.isHidden = tagType == .select
    }
    // MARK: - Layout
    public override func layoutSubviews() {
        button.sizeToFit()
        if tagType == .select {
            button.frame.origin = CGPoint(x: 20, y: 20)
        } else {
            button.frame.origin = CGPoint(x: 44 , y: 20)
            cancelButton.frame.size = CGSize(width: 20, height: 20)
            cancelButton.frame.origin = CGPoint(x: 20 , y: 20 + button.frame.height/2 - 10)
        }
        frame.size = CGSize(width: button.frame.maxX + 20, height: button.frame.maxY + 20)
    }
    
    func updateAll() {
        cancelButton.isHidden = tagType == .select
        updateTitlesColorsAndFontsDueToSelection()
        updateBorder()
        button.sizeToFit()
        layoutIfNeeded()
        invalidateIntrinsicContentSize()
    }
    
    func updateTitlesColorsAndFontsDueToSelection() {
        backgroundColor = isSelected ? tagMainBackColor : tagSecondBackColor
        let textColor = isSelected ? tagMainTextColor : tagSecondTextColor
        var attributes: [String: Any] = [:]
        attributes[NSFontAttributeName] = tagFont
        attributes[NSForegroundColorAttributeName] = textColor
        button.setAttributedTitle(NSAttributedString(string: tagTitle, attributes: attributes), for: .normal)
    }
    
    func updateBorder() {
        layer.borderWidth = tagBorderWidth
        layer.borderColor = tagBorderColor
        layer.cornerRadius = bounds.height * tagCornerRadiusToHeightRatio
    }
    
    // MARK: - User interaction
    func tapped(){
        print("tapped")
        delegate?.tagClicked(self)
    }
    func cancelled() {
        print("cancelled")
        delegate?.tagCancelled(self)
    }
}


extension UIColor {
    
    func lighter(by percentage: CGFloat = 30.0) -> UIColor? {
        return self.adjust(by: abs(percentage) )
    }
    
    func darker(by percentage: CGFloat = 30.0) -> UIColor? {
        return self.adjust(by: -1 * abs(percentage) )
    }
    
    func adjust(by percentage: CGFloat = 30.0) -> UIColor? {
        var r: CGFloat = 0
        var g: CGFloat = 0
        var b: CGFloat = 0
        var a: CGFloat = 0
        
        if(self.getRed(&r, green: &g, blue: &b, alpha: &a)){
            return UIColor(red: min(r + percentage/100, 1.0),
                           green: min(g + percentage/100, 1.0),
                           blue: min(b + percentage/100, 1.0),
                           alpha: a)
        }else{
            return nil
        }
    }
}


HTagView/Classes/HTag.swift
//
//  HTag.swift
//  Pods
//
//  Created by Chang, Hao on 12/12/2016.
//
//

import UIKit

protocol HTagDelegate: class {
    func tagClicked(_ sender: HTag)
    func tagCancelled(_ sender: HTag)
}

public class HTag: UIView {
    weak var delegate: HTagDelegate?
    
    // MARK: - HTag Configuration
    /**
     Type of tag
     */
    var tagType: HTagType = .cancel { didSet { updateAll() } }
    /**
     Title of tag
     */
    var tagTitle: String = "" { didSet { updateAll() } }
    /**
     Main background color of tags
     */
    var tagMainBackColor : UIColor = UIColor.white { didSet { updateTitlesColorsAndFontsDueToSelection() } }
    /**
     Main text color of tags
     */
    var tagMainTextColor : UIColor = UIColor.black { didSet { updateTitlesColorsAndFontsDueToSelection() } }
    /**
     Secondary background color of tags
     */
    var tagSecondBackColor : UIColor = UIColor.gray { didSet { updateTitlesColorsAndFontsDueToSelection() } }
    /**
     Secondary text color of tags
     */
    var tagSecondTextColor : UIColor = UIColor.white { didSet { updateTitlesColorsAndFontsDueToSelection() } }
    /**
     The border width to height ratio of HTags.
     */
    var tagBorderWidth :CGFloat = 1 { didSet { updateBorder() } }
    /**
     The border color to height ratio of HTags.
     */
    var tagBorderColor :CGColor? = UIColor.darkGray.cgColor { didSet { updateBorder() } }
    
    /**
     The corner radius to height ratio of HTags.
     */
    var tagCornerRadiusToHeightRatio: CGFloat = 0.2 { didSet { updateBorder() } }
    /**
     The content EdgeInsets of HTags, which would automatically adjust the position in `.cancel` type.
     On the other word, usages are the same in both types.
     */
    var tagContentEdgeInsets: UIEdgeInsets = UIEdgeInsets() { didSet { updateAll()} }
    /**
     The distance between cancel icon and text
     */
    var tagCancelIconRightMargin: CGFloat = 4 { didSet { updateAll()} }
    /**
     The Font of HTags.
     */
    var tagFont: UIFont = UIFont.systemFont(ofSize: 17) { didSet { updateAll() } }
    
    /**
     Specified Width of Tag
     */
    var tagSpecifiedWidth: CGFloat? = nil { didSet { layoutSubviews() } }
    /**
     Maximum Width of Tag
     */
    var tagMaximumWidth: CGFloat? = nil { didSet { layoutSubviews() } }
    // MARK: - status
    private(set) var isSelected: Bool = false
    
    
    // MARK: - Subviews
    var button: UIButton = UIButton(type: .system)
    var cancelButton: UIButton = UIButton(type: .system)
    
    // MARK: - Init
    override init(frame: CGRect){
        super.init(frame: frame)
        configure()
    }
    
    required public init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
        configure()
    }
    
    deinit {
        button.removeObserver(self, forKeyPath: "highlighted")
    }
    
    // MARK: - Configure
    func configure(){
        clipsToBounds = true
        self.addGestureRecognizer(UITapGestureRecognizer(target: self, action: #selector(tapped)))
        setupButton()
        setupCancelButton()
        updateAll()
    }
    
    // MARK: - button
    func setupButton() {
        addSubview(button)
        button.addTarget(self, action: #selector(tapped), for: .touchUpInside)
        button.addObserver(self, forKeyPath: "highlighted", options: .new, context: nil)
    }
    
    func setHighlight(_ highlighted: Bool) {
        let color = isSelected ? tagMainBackColor : tagSecondBackColor
        backgroundColor = highlighted ? color.darker() : color
    }
    func setSelected(_ selected: Bool) {
        isSelected = selected
        updateTitlesColorsAndFontsDueToSelection()
    }
    public override func observeValue(forKeyPath keyPath: String?, of object: Any?, change: [NSKeyValueChangeKey : Any]?, context: UnsafeMutableRawPointer?) {
        
        guard let keyPath = keyPath else {
            return
        }
        switch keyPath {
        case "highlighted":
            guard let highlighted = change?[.newKey] as? Bool else  {
                break
            }
            setHighlight(highlighted)
        default:
            break
        }
    }
    
    // MARK: - cancel button
    func setupCancelButton() {
        cancelButton.setImage(UIImage(named: "close_small", in: Bundle(for: self.classForCoder), compatibleWith: nil)?.withRenderingMode(.alwaysTemplate), for: UIControlState())
        cancelButton.addTarget(self, action: #selector(cancelled), for: .touchUpInside)
        addSubview(cancelButton)
        cancelButton.isHidden = tagType == .select
    }
    // MARK: - Layout
    public override func layoutSubviews() {
        button.sizeToFit()
        // set origin
        if tagType == .select {
            button.frame.origin = CGPoint(x: tagContentEdgeInsets.left, y: tagContentEdgeInsets.top)
        } else {
            let cancelButtonSide: CGFloat = 16
            cancelButton.frame.size = CGSize(width: cancelButtonSide, height: cancelButtonSide)
            cancelButton.frame.origin = CGPoint(x: tagContentEdgeInsets.left ,
                                                y: tagContentEdgeInsets.top + button.frame.height/2 - cancelButtonSide/2)
            button.frame.origin = CGPoint(x: tagContentEdgeInsets.left + cancelButtonSide + tagCancelIconRightMargin ,
                                          y: tagContentEdgeInsets.top)
        }
        
        // set size
        let tagHeight =  button.frame.maxY + tagContentEdgeInsets.bottom
        var tagWidth: CGFloat = 0
        if let tagSpecifiedWidth = tagSpecifiedWidth {
            tagWidth = tagSpecifiedWidth
        } else {
            tagWidth = button.frame.maxX + tagContentEdgeInsets.right
        }
        
        var shouldSetButtonWidth: Bool = false
        if let maxWidth = tagMaximumWidth,
            maxWidth < tagWidth {
            tagWidth = maxWidth
            shouldSetButtonWidth = true
        } else if tagSpecifiedWidth != nil {
            shouldSetButtonWidth = true
        }
        frame.size = CGSize(width: tagWidth, height: tagHeight)
        
        if shouldSetButtonWidth {
            button.frame.size.width = tagWidth
                - cancelButton.frame.width - tagCancelIconRightMargin
                - tagContentEdgeInsets.left - tagContentEdgeInsets.right
        }
    }
    
    func updateAll() {
        cancelButton.isHidden = tagType == .select
        updateTitlesColorsAndFontsDueToSelection()
        updateBorder()
        button.sizeToFit()
        layoutIfNeeded()
        invalidateIntrinsicContentSize()
    }
    
    func updateTitlesColorsAndFontsDueToSelection() {
        backgroundColor = isSelected ? tagMainBackColor : tagSecondBackColor
        let textColor = isSelected ? tagMainTextColor : tagSecondTextColor
        var attributes: [String: Any] = [:]
        attributes[NSFontAttributeName] = tagFont
        attributes[NSForegroundColorAttributeName] = textColor
        button.setAttributedTitle(NSAttributedString(string: tagTitle, attributes: attributes), for: .normal)
        if tagType == .cancel {
            cancelButton.tintColor = textColor
        }
    }
    
    func updateBorder() {
        layer.borderWidth = tagBorderWidth
        layer.borderColor = tagBorderColor
        layer.cornerRadius = bounds.height * tagCornerRadiusToHeightRatio
    }
    
    // MARK: - User interaction
    func tapped(){
        delegate?.tagClicked(self)
    }
    func cancelled() {
        delegate?.tagCancelled(self)
    }
}


extension UIColor {
    
    func lighter(by percentage: CGFloat = 30.0) -> UIColor? {
        return self.adjust(by: abs(percentage) )
    }
    
    func darker(by percentage: CGFloat = 30.0) -> UIColor? {
        return self.adjust(by: -1 * abs(percentage) )
    }
    
    func adjust(by percentage: CGFloat = 30.0) -> UIColor? {
        var r: CGFloat = 0
        var g: CGFloat = 0
        var b: CGFloat = 0
        var a: CGFloat = 0
        
        if(self.getRed(&r, green: &g, blue: &b, alpha: &a)){
            return UIColor(red: min(r + percentage/100, 1.0),
                           green: min(g + percentage/100, 1.0),
                           blue: min(b + percentage/100, 1.0),
                           alpha: a)
        }else{
            return nil
        }
    }
}


HTagView/Classes/HTagView.swift
//
//  HTagView.swift
//  Pods
//
//  Created by Chang, Hao on 5/16/16.
//
//

import UIKit

/**
 HTagViewDelegate is a protocol to implement for responding to user interactions to the HTagView.
 */
@objc
public protocol HTagViewDelegate: class {
    /**
     Called when user did cancel tag at index
     */
    @objc optional func tagView(_ tagView: HTagView, didCancelTagAtIndex index: Int)
    @objc optional func tagView(_ tagView: HTagView, tagSelectionDidChange selectedIndices: [Int])
}

/**
 HTagViewDataSource is a protocol to implement for data source of the HTagView.
 */
public protocol HTagViewDataSource: class {
    func numberOfTags(_ tagView: HTagView) -> Int
    func tagView(_ tagView: HTagView, titleOfTagAtIndex index: Int) -> String
    func tagView(_ tagView: HTagView, tagTypeAtIndex index: Int) -> HTagType
    func tagView(_ tagView: HTagView, tagWidthAtIndex index: Int) -> CGFloat
}

public extension HTagViewDataSource {
    func tagView(_ tagView: HTagView, tagWidthAtIndex index: Int) -> CGFloat {
        return .HTagAutoWidth
    }
}

/**
 HTag comes with two types, `.cancel` and `.select`.
 */
public enum HTagType{
    case cancel, select
}

/**
 HTagView is customized tag view sublassing UIView where tag could be either with cancel button or seletable.
 */
@IBDesignable
open class HTagView: UIView {
    
    // MARK: - DataSource
    /**
     HTagViewDataSource
     */
    open weak var dataSource : HTagViewDataSource?{
        didSet{
            reloadData()
        }
    }
    
    // MARK: - Delegate
    /**
     HTagViewDelegate
     */
    open weak var delegate : HTagViewDelegate?
    
    // MARK: - HTagView Configuration
    /**
     HTagView multiselect
     */
    @IBInspectable
    open var multiselect : Bool = true
    /**
     HTagView margin
     */
    @IBInspectable
    open var marg : CGFloat = 20
    /**
     Distance between tags in the same line
     */
    @IBInspectable
    open var btwTags : CGFloat = 8
    /**
     Distance between lines
     */
    @IBInspectable
    open var btwLines : CGFloat = 8
    
    // MARK: - HTag Configuration
    /**
     Main background color of tags
     */
    @IBInspectable
    open var tagMainBackColor : UIColor = UIColor(colorLiteralRed: 100/255, green: 200/255, blue: 205/255, alpha: 1) {
        didSet {
            tags.forEach {
                $0.tagMainBackColor = tagMainBackColor
            }
        }
    }
    /**
     Main text color of tags
     */
    @IBInspectable
    open var tagMainTextColor : UIColor = UIColor.white {
        didSet {
            tags.forEach {
                $0.tagMainTextColor = tagMainTextColor
            }
        }
    }
    /**
     Secondary background color of tags
     */
    @IBInspectable
    open var tagSecondBackColor : UIColor = UIColor.lightGray {
        didSet {
            tags.forEach {
                $0.tagSecondBackColor = tagSecondBackColor
            }
        }
    }
    /**
     Secondary text color of tags
     */
    @IBInspectable
    open var tagSecondTextColor : UIColor = UIColor.darkText {
        didSet {
            tags.forEach {
                $0.tagSecondTextColor = tagSecondTextColor
            }
        }
    }
    /**
     The border width to height ratio of HTags.
     */
    @IBInspectable
    open var tagBorderWidth :CGFloat = CGFloat(0){
        didSet{
            tags.forEach {
                $0.tagBorderWidth = tagBorderWidth
            }
            for tag in tags{
                tag.layer.borderWidth = tagBorderWidth
            }
            layoutIfNeeded()
        }
    }
    /**
     The border color to height ratio of HTags.
     */
    @IBInspectable
    open var tagBorderColor :CGColor? = nil{
        didSet{
            tags.forEach {
                $0.tagBorderColor = tagBorderColor
            }
            for tag in tags{
                tag.layer.borderColor = tagBorderColor
            }
            layoutIfNeeded()
        }
    }
    /**
     Maximum Width of Tag
     */
    open var tagMaximumWidth: CGFloat? = .HTagAutoMaximumWidth {
        didSet {
            tags.forEach {
                $0.tagMaximumWidth = tagMaximumWidth
            }
            layoutIfNeeded()
        }
    }
    /**
     The corner radius to height ratio of HTags.
     */
    @IBInspectable
    open var tagCornerRadiusToHeightRatio :CGFloat = CGFloat(0.2){
        didSet{
            tags.forEach {
                $0.tagCornerRadiusToHeightRatio = tagCornerRadiusToHeightRatio
            }
            for tag in tags{
                tag.layer.cornerRadius = tag.frame.height * tagCornerRadiusToHeightRatio
            }
            layoutIfNeeded()
        }
    }
    /**
     The content EdgeInsets of HTags, which would automatically adjust the position in `.cancel` type.
     On the other word, usages are the same in both types.
     */
    @IBInspectable
    open var tagContentEdgeInsets: UIEdgeInsets = UIEdgeInsets(top: 0, left: 8, bottom: 0, right: 8) {
        didSet{
            for tag in tags{
                tag.tagContentEdgeInsets = tagContentEdgeInsets
            }
        }
    }
    /**
     The distance between cancel icon and text
     */
    @IBInspectable
    open var tagCancelIconRightMargin: CGFloat = 4 {
        didSet{
            for tag in tags{
                tag.tagCancelIconRightMargin = tagCancelIconRightMargin
            }
        }
    }
    /**
     The Font of HTags.
     */
    @IBInspectable
    open var tagFont : UIFont = UIFont.systemFont(ofSize: 17) {
        didSet{
            for tag in tags {
                tag.tagFont = tagFont
            }
            layoutIfNeeded()
            invalidateIntrinsicContentSize()
        }
    }
    /**
     Indices of selected tags.
     */
    open var selectedIndices: [Int]{
        get{
            var selectedIndexes = [Int]()
            for (index, tag) in tags.enumerated(){
                if tag.tagType == .select && tag.isSelected{
                    selectedIndexes.append(index)
                }
            }
            return selectedIndexes
        }
    }
    
    var tags: [HTag] = []
    
    // MARK: - init
    override public init(frame: CGRect){
        super.init(frame: frame)
        configure()
        reloadData(false)
    }
    
    required public init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
        configure()
        reloadData(false)
    }
    
    func configure(){
    }
    
    // MARK: - reload
    open func reloadData(_ preserveSelectionState: Bool = true){
        guard let dataSource = dataSource else {
            return
        }
        
        let selection = selectedIndices
        
        for tag in tags {
            tag.removeFromSuperview()
        }
        tags = []
        for index in  0 ..< dataSource.numberOfTags(self) {
            let tag = HTag()
            tag.delegate = self
            tag.tagType = dataSource.tagView(self, tagTypeAtIndex: index)
            if tag.tagType == .select {
                tag.setSelected(preserveSelectionState ? selection.contains(index) : false)
            }
            tag.tagContentEdgeInsets = tagContentEdgeInsets
            tag.tagFont = tagFont
            tag.tagMainBackColor = tagMainBackColor
            tag.tagSecondBackColor = tagSecondBackColor
            tag.tagMainTextColor = tagMainTextColor
            tag.tagSecondTextColor = tagSecondTextColor
            tag.tagBorderColor = tagBorderColor
            tag.tagBorderWidth = tagBorderWidth
            tag.tagCornerRadiusToHeightRatio = tagCornerRadiusToHeightRatio
            tag.tagCancelIconRightMargin = tagCancelIconRightMargin
            tag.tagMaximumWidth = tagMaximumWidth
            tag.tagTitle = dataSource.tagView(self, titleOfTagAtIndex: index)
            addSubview(tag)
            tags.append(tag)
        }
        
        layoutSubviews()
        invalidateIntrinsicContentSize()
    }
    
    // MARK: - Subclassing UIView
    override open func layoutSubviews() {
        guard let dataSource = dataSource else {
            return
        }
        
        if dataSource.numberOfTags(self) == 0 {
            self.frame.size = CGSize(width: self.frame.width, height: 0)
        }else{
            var x = marg
            var y = marg
            for index in 0..<tags.count{
                if dataSource.tagView(self, tagWidthAtIndex: index) != .HTagAutoWidth {
                    tags[index].tagSpecifiedWidth = dataSource.tagView(self, tagWidthAtIndex: index)
                } else {
                    tags[index].tagSpecifiedWidth = nil
                }
                if tagMaximumWidth == .HTagAutoMaximumWidth {
                    tags[index].tagMaximumWidth = frame.width - 2 * marg
                }
                if tags[index].frame.width + x > frame.width - marg{
                    y += tags[index].frame.height + btwLines
                    x = marg
                }
                
                tags[index].frame.origin = CGPoint(x: x, y: y)
                x += tags[index].frame.width + btwTags
            }
            self.frame.size = CGSize(width: self.frame.width, height: y + (tags.last?.frame.height ?? 0) + marg )
        }
    }
    
    override open var intrinsicContentSize : CGSize {
        if tags.count == 0{
            return CGSize(width: UIViewNoIntrinsicMetric, height: 0)
        }else{
            let height = (tags.last?.frame.origin.y ?? 0) + (tags.last?.frame.height ?? 0) + marg
            return CGSize(width: UIViewNoIntrinsicMetric, height: height )
        }
    }
    
    // MARK: - Manipulate Tags
    /**
     Select on tag with titles in `.select` type HTagView. The delegate method `tagView(_:tagSelectionDidChange:)` will be called.
     */
    open func selectTagAtIndex(_ index: Int){
        for (i, tag) in tags.enumerated() {
            guard let type = dataSource?.tagView(self, tagTypeAtIndex: i), type == .select else {
                continue
            }
            
            if i == index {
                tag.setSelected(true)
            } else if !multiselect {
                tag.setSelected(false)
            }
        }
    }
    /**
     Deselect on tag with titles in `.select` type HTagView. The delegate method `tagView(_:tagSelectionDidChange:)` will be called.
     */
    open func deselectTagAtIndex(_ index: Int){
        guard let type = dataSource?.tagView(self, tagTypeAtIndex: index) , type == .select else {
            return
        }
        tags[index].setSelected(false)
    }
    
}

extension HTagView: HTagDelegate {
    func tagClicked(_ sender: HTag) {
        guard let index = tags.index(of: sender) else{
            return
        }
        if dataSource?.tagView(self, tagTypeAtIndex: index) == .select{
            if sender.isSelected {
                deselectTagAtIndex(index)
            }else{
                selectTagAtIndex(index)
            }
        }
        delegate?.tagView?(self, tagSelectionDidChange: selectedIndices)
    }
    func tagCancelled(_ sender: HTag) {
        guard let index = tags.index(of: sender) else{
            return
        }
        delegate?.tagView?(self, didCancelTagAtIndex: index)
    }
}


Example/HTagView/ViewController.swift
//
//  ViewController.swift
//  HTagView
//
//  Created by Chang, Hao on 05/16/2016.
//  Copyright (c) 2016 Chang, Hao. All rights reserved.
//

import UIKit
import HTagView

class ViewController: UIViewController, HTagViewDelegate, HTagViewDataSource {

    @IBOutlet weak var tagView1: HTagView!
    @IBOutlet weak var tagView2: HTagView!
    override func viewDidLoad() {
        super.viewDidLoad()
        
        tagView1.delegate = self
        tagView1.dataSource = self
        tagView1.multiselect = false
        tagView1.marg = 20
        tagView1.btwTags = 20
        tagView1.btwLines = 20
        tagView1.tagFont = UIFont.systemFont(ofSize: 15)
        tagView1.tagMainBackColor = UIColor(red: 121/255, green: 196/255, blue: 1, alpha: 1)
        tagView1.tagMainTextColor = UIColor.white
        tagView1.tagSecondBackColor = UIColor.lightGray
        tagView1.tagSecondTextColor = UIColor.darkText
        
        
        tagView2.delegate = self
        tagView2.dataSource = self
        tagView2.marg = 20
        tagView2.btwTags = 20
        tagView2.btwLines = 20
        tagView2.tagFont = UIFont.systemFont(ofSize: 15)
        tagView2.tagMainBackColor = UIColor(red: 1, green: 130/255, blue: 103/255, alpha: 1)
        tagView2.tagSecondBackColor = UIColor.lightGray
        tagView2.tagSecondTextColor = UIColor.darkText
        tagView2.tagContentEdgeInsets = UIEdgeInsets(top: 10, left: 20, bottom: 10, right: 20)
        tagView2.tagMaximumWidth = .HTagAutoMaximumWidth
        tagView2.tagBorderColor = UIColor.black.cgColor
        tagView2.tagBorderWidth = 2
        
        
        tagView2.selectTagAtIndex(6)
        tagView1.selectTagAtIndex(3)
        
        tagView1.reloadData()
        tagView2.reloadData()
        
    }
    
    // MARK: - Data
    let tagView1_data = ["Hey!","This","is","a","HTagView."]
    var tagView2_data = ["Hey!","This","is","a","HTagView", "as", "well.", "WEOPIkl,.cviwoipaeai;kdxioaw389WOIERJAAW;EOIFJAOIEJWFAWEIOFPJIPOJ"]
    
    // MARK: - HTagViewDataSource
    func numberOfTags(_ tagView: HTagView) -> Int {
        switch tagView {
        case tagView1:
            return tagView1_data.count
        case tagView2:
            return tagView2_data.count
        default:
            return 0
        }
    }
    
    func tagView(_ tagView: HTagView, titleOfTagAtIndex index: Int) -> String {
        switch tagView {
        case tagView1:
            return tagView1_data[index]
        case tagView2:
            return tagView2_data[index]
        default:
            return "???"
        }
    }
    
    func tagView(_ tagView: HTagView, tagTypeAtIndex index: Int) -> HTagType {
        switch tagView {
        case tagView1:
            return index > 0 ? .select : .cancel
        case tagView2:
            return index > 3 ? .select : .cancel
        default:
            return .select
        }
    }
    
    func tagView(_ tagView: HTagView, tagWidthAtIndex index: Int) -> CGFloat {
        return .HTagAutoWidth
//        return 150
    }
    
    // MARK: - HTagViewDelegate
    func tagView(_ tagView: HTagView, tagSelectionDidChange selectedIndices: [Int]) {
        print("tag with indices \(selectedIndices) are selected")
    }
    func tagView(_ tagView: HTagView, didCancelTagAtIndex index: Int) {
        print("tag with index: '\(index)' has to be removed from tagView")
        tagView2_data.remove(at: index)
        tagView.reloadData()
    }
    

}

